CMR/python/docit.py

```diff
@@ -0,0 +1,387 @@
+                    if func_name in config:
+                        out.append(config.get(func_name, '') + '\n')
+                data = json.loads(item)
+
+                if 'from' in data:
+                    fix_local_from(module, data)
+                    find_froms(func_db, out, data)
+                else:
+                    txt = document_it_tag_to_str(item)
+                    out.append ("* {}".format( txt ))
+                first = False
+    if not first:
+        out.append('')
+
+def handle_builtins(out, obj_info, config:dict):
+    """
+    Handle the built-in use case
+    Parameters:
+        out: list of strings to be sent through print() as the markdown content
+        obj_info: (object type, object)
+        config: configuration dictionary, look for 'verbose'
+    """
+    obj_type = obj_info[0]
+    obj = obj_info[1]
+    if config.get('verbose', False) and obj.__doc__ is not None:
+        out.append("* Global Dictionary - {}".format(obj_type))
+        out.append (obj.__doc__)
+
+
+# ##############################################################################
+#mark - Public Functions
+
+def pre_tree(func_db, module, config=None, cache=None):
+    """
+    Scan the modules ahead of time and make a map of all the documents-its for
+    future use. Must be called before tree(). This database of functions will
+    then be called by tree() to query and walk the 'froms'
+    """
+    config = config if isinstance(config, dict) else {}
+    cache = cache if isinstance(cache, dict) else {}
+    if module.__name__ in cache:
+        # Been here before, nothing to do
+        return
+    cache[module.__name__] = {'processed':True}
+    for obj_type, obj in inspect.getmembers(module):
+        if inspect.ismodule(obj):
+            if obj.__package__.startswith(module.__package__):
+                pre_tree(func_db, obj, config, cache)
+        elif inspect.isfunction(obj):
+            comments = inspect.getcomments(obj)
+            if comments is not None and len(comments) > 0:
+                for raw_item in comments.strip().split("\n"):
+                    item = raw_item.strip()
+                    if item.startswith(DOCUMENT_IT):
+                        key = module.__name__ + "." + obj_type
+                        value = json.loads(item[len(DOCUMENT_IT):])
+                        fix_local_from(module, value)
+                        add_to_tree_database(func_db, key, value)
+
+def tree(report, func_db, module, config=None, cache=None):
+    """
+    Walk the module tree and create the mark down output
+    Parameters:
+        report: list of strings, output for print
+        func_db: dictionary of known functions
+        module: current module being processed
+        config: settings
+        cache: record of which modules have been processed
+    """
+    cache = cache if isinstance(cache, dict) else {}
+    config = config if isinstance(config, dict) else {}
```

Comment:
same comment about modularization of line 269. 

I notice that line 268 is pretty similar: ```x = x if isinstance(x, dict) else {}```. If this is a big pattern in this project then maybe it should be modularized, too.

Reply:
that's funny, I just un-modularized it

Reply:
make sure the function is generic enough to handle both cache and config

Reply:
wrote a new function called always().